### ISSUE TITLE
allow karpenter to create nodes with non-well-known kubernetes labels

### ISSUE THREAD

**Issue by elmiko** [kubernetes-sigs/karpenter#1046]
### Description

**What problem are you trying to solve?**

As an openshift user, i require new nodes to take the label `node-role.kubernetes.io/worker`. Although this label uses the `kubernetes.io` prefix, it is not an [official well-known label](https://kubernetes.io/docs/reference/labels-annotations-taints/). Currently, karpenter does not allow NodePools to be created when a label with the `kubernetes.io` prefix is used on a non-well-known label for new node creation, or when checking predicates.

**How important is this feature to you?**

This is fairly important for openshift as the label is used frequently. It is possible to work around this, but for long term usage on openshift a permanent solution would be better.

This issue was [discussed at the Feb. 26 2024 community meeting](https://docs.google.com/document/d/18BT0AIMugpNpiSPJNlcAL2rv69yAE6Z06gUVj7v_clg/edit#bookmark=id.9809ti3z75rr).

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
Looks like there are a couple other issues that were asking for similar functionality here: https://github.com/aws/karpenter-provider-aws/issues/1941 and https://github.com/kubernetes-sigs/karpenter/issues/660

**Comment by jonathan-innis** [comment]
What's the security implication of allowing users to set this custom label? From a little sleuthing, it doesn't seem to be any different than the `node-restriction.kubernetes.io` label that we already support today.

**Comment by sftim** [comment]
As Karpenter is a Kubernetes project, I'm wary of even indirectly encouraging use of unregistered labels.

If `node-role.kubernetes.io/worker` should become an official label, let's get that registered first - it's not a particularly difficult process. However, registration **first**, then adoption.

“OpenShift uses it” does not count as grounds to add support into Kubernetes code.

--- 

OpenShift folks, if you want a label to mark nodes as OpenShift workers, you can make one, eg `openshift.example/worker: true`

**Comment by elmiko** [comment]
> As Karpenter is a Kubernetes project, I'm wary of even indirectly encouraging use of unregistered labels.

this makes me wonder if the kubelet should be creating warnings/errors if the user attempts to use an unregistered well-known label?

karpenter is preventing something that the kubelet allows.

> If node-role.kubernetes.io/worker should become an official label, let's get that registered first - it's not a particularly difficult process. However, registration first, then adoption.

totally agreement from me, but this is historical as well. openshift has been using the label for a long time, i only mention this to help add context around the effort to change the value.

> “OpenShift uses it” does not count as grounds to add support into Kubernetes code.

agreed, but let's take openshift out of the equation. as a user, i would like karpenter to create nodes with unregistered labels from the kubernetes.io prefix, i can do this today on the kubelet command line and i would like karpenter to do the same for me.

i think if this sounds like an unreasonable request at the karpenter level, then we should consider making the kubelet produce warnings when this happens because we are now asking karpenter to gate activity that the kubelet allows.

> OpenShift folks, if you want a label to mark nodes as OpenShift workers, you can make one, eg openshift.example/worker: true

no objection from me, but i think this is orthogonal to the true issue here. i only mentioned openshift to help illuminate the use case.

> So long as we leave the kubelet to set that label, we're fine. If Karpenter starts setting a label on a node, it could provide a way to bypass restrictions on what labels nodes can have (although, letting someone write to your NodePools is already bad practice).

i think this is question is important, if karpenter is adding the label after node creation, then i totally agree with it preventing unregistered labels from being added. but if karpenter is passing the labels down to the kubelet configuration, then i think we should reconsider.

thanks for the thoughtful discussion @sftim and @jonathan-innis

**Comment by elmiko** [comment]
one additional thought on this, even if karpenter cannot create nodes with the unregistered labels, it would be nice to able to have it know when a node will be created with that label so that it can properly predict which nodes should be created.

essentially, i am fine with karpenter not placing the unregistred label on the node, but i should be able to tell karpenter "hey if you see these pods, put them on nodes will this label, and oh btw these node classes will have that label".

**Comment by sftim** [comment]
We have code in Kubernetes to [prevent](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) the node setting disallowed labels. You can use Karpenter to bypass that; it's your cluster.

However, should we **encourage** it? I say we should not. Upstream Kubernetes and in-project code should abide by our architecture standards, even if individual end users intent to disregard them.

**Comment by elmiko** [comment]
> However, should we encourage it? I say we should not.

agreed

> TL;DR: offering an in-project security footgun feels like an obviously bad choice.

just to be clear, do you consider the footgun to be: karpenter adding disallowed labels after the node is created, karpenter adding disallowed labels to the command line flag list for kubelet, or both?

**Comment by elmiko** [comment]
reading this a little deeper, https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction

i have a question about this interaction with the admission controllers, if a user has disabled the admission controller for node restriction, will that change be reflected in karpenter's restriction as well?

it seems like we would want karpenter to reflect what the admission controllers are doing, so if an update to a node object get rejected wouldn't we want to surface that error to the user?

**Comment by sftim** [comment]
I suppose it's an outright footgun if a compromised node can, in any way, set labels on itself it shouldn't be able to, or cause those to be set.

I'm also opposed to tacitly encouraging people to use labels that Kubernetes doesn't recognize. We shouldn't staff work that encourages doing more of the wrong thing.

**Comment by sftim** [comment]
> i have a question about this interaction with the admission controllers, if a user has disabled the admission controller for node restriction, will that change be reflected in karpenter's restriction as well?

No; in general, Karpenter cannot know what admission controllers are in use.

**Comment by elmiko** [comment]
> No; in general, Karpenter cannot know what admission controllers are in use.

perhaps i phrased that poorly, i'll try to restate.

imagine a user has the noderestriction admission controller disabled. the user configures karpenter to make changes that would otherwise be prevented by the admission controller. karpenter refuses to allow the changes because it has duplicated the default rules for the noderestriction admission cotroller. should karpenter be the component responsible for rejecting the configuration?

i'm wondering if karpenter should be returning errors from updating nodes/pods instead of interposing itself.

> We shouldn't staff work that encourages doing more of the wrong thing.

agreed, and this makes a lot of sense to me.

**Comment by sftim** [comment]
> should karpenter be the component responsible for rejecting the configuration?

I think so, because it's much easier to maintain an important security guarantee this way.
It sounds like you think Karpenter is an admission controller; it's not.

**Comment by sftim** [comment]
> i don't think i do, but i am trying to reconcile why the rules for the unregistered labels are encoded/validated in karpenter instead of deferring to the apiserver to return an error. (perhaps i am not understanding this flow properly).

The API server's restrictions (from the node restriction admission controller) apply exclusively to label changes made by nodes. Karpenter is not a node, so those restrictions wouldn't be enforced.

**Comment by elmiko** [comment]
happy to see this considered for inclusion, but given the conversation between @sftim and myself, i'm not sure this is a good idea to carry forward. it seems like a possible point for a security escalation.

**Comment by sftim** [comment]
I think we should ask the OpenShift folks to consider not using the wrong label, and leave it at that. OpenShift is entirely welcome to use _OpenShift_ labels for node role, it just shouldn't bypass K8s processes for registering labels.

**Comment by sftim** [comment]
For people using Karpenter who want to work around this, they could use a different label (eg `internal-node-role: worker`) and a contrib controller that copies `internal-node-role` to whatever other label they want set. But we shouldn't make that contrib code part of Kubernetes either, because it'd be furthering a mistake.

**Comment by diranged** [comment]
So given that Karpenter already applies Taints to Node objects upon creation, I can't see why it would be any different to allow Karpenter to apply labels (including `node-role.kubernetes.io/<role>`).

**Comment by sftim** [comment]
Kubernetes enforces the rules here; a node autoscaler, as a client of the
Kubernetes API, isn't able by itself to bypass the security rule.Message
ID: ***@***.***>

>

**Comment by diranged** [comment]
> Kubernetes enforces the rules here; a node autoscaler, as a client of the Kubernetes API, isn't able by itself to bypass the security rule.Message ID: ***@***.***>
> […](#)

I don't believe that is true at all. For example, the Spot.io Ocean Cluster Controller absolutely labels our nodes with the `node-role` label for us on startup. The only rule here is that the _node itself_ is unable to apply that label.

**Comment by sftim** [comment]
The label node-role is not reserved by Kubernetes (and as a cluster-private
label, never would be).

There is no restriction on labelling nodes with private labels, only on
adding registered public labels that belong to Kubernetes.

**Comment by sftim** [comment]
Should node autoscalers try to set unofficial labels in the Kubernetes
namespace? It's possible to code up, but I (still) don't recommend it; in
fact, I actively oppose making that change to Karpenter.
See
https://github.com/kubernetes-sigs/karpenter/issues/1046#issuecomment-2237753906

**Comment by diranged** [comment]
> Should node autoscalers try to set unofficial labels in the Kubernetes namespace? It's possible to code up, but I (still) don't recommend it; in fact, I actively oppose making that change to Karpenter

I should fix my comment - I  want to use the `kubernetes.io/role=xxx` label - specifically defined here https://github.com/kubernetes/kubernetes/blob/v1.30.3/staging/src/k8s.io/kubectl/pkg/describe/interface.go#L33-L34 as:
```go
// NodeLabelRole specifies the role of a node
NodeLabelRole = "kubernetes.io/role"
```
Can you elaborate on _why_ you oppose Karpenter managing this label? I can't find anything in the Kubernetes code or documentation that indicates we _should't_ set this label.

**Comment by elmiko** [comment]
> I think we should ask the OpenShift folks to consider not using the wrong label, and leave it at that. OpenShift is entirely welcome to use _OpenShift_ labels for node role, it just shouldn't bypass K8s processes for registering labels.

i think this is sound advice and something i have advocated for internally at red hat. i only brought up openshift in my issue as it pertains to my specific problem here. i am not attempting to suggest that we need to change kubernetes core behaviors based on my needs, i only added it to help with context.

i appreciate the extended discussions from @sftim and the explanation for the security-related part of this issue.

my confusion is more around karpenter attempting to block everything in the `kuberenetes.io` prefixed labels, i think this is similar to what @diranged is also pointing out. on this point, i'm fine to accept the community's decision on what karpenter should or should not allow, i was mostly looking to have an open discussion about it.

**Comment by sftim** [comment]
Maybe this is a missing detail: Karpenter creates nodes that have labels
by, AIUI, creating nodes that will run a kubelet that then self-labels.

We could enhance Karpenter to also use a different mechanism, setting
labels directly on Nodes based on the owning NodeClaim. I don't see a good
case for adding complexity here, but a case could be made.

**Comment by diranged** [comment]
> creating nodes that will run a kubelet that then self-labels

@sftim,
  So that is true - if you are using an amiFamily like `Bottlerocket` where Karpenter creates/merges the userData... however, we actually were running in `Custom` mode, where Karpenter is not mutating or managing the userData at all, and it still sets the labels on the nodes. Karpenter is (I think) pre-creating the `Node` object and populating the labels that way..  which would allow it to set the `kubernetes.io/role` label just fine.

**Comment by sftim** [comment]
Also see https://github.com/kubernetes/website/issues/48803

**Comment by sftim** [comment]
Should _Karpenter_ be allowed to set node role labels on behalf of the nodes it deploys? That's still a tricky one to answer.

**Comment by elmiko** [comment]
personally, i would love to see karpenter open up it's allowed labels here, for reading and writing. currently there are some issues where karpenter uses the same detection logic to determine if the labels can be used for selection (not even for writing).

i totally get the reasoning for why kubelets are restricted in this manner, but i think other controllers should be able to label after the node has been created. the only thing i'm not sure about here is whether karpenter labels after the node is created or if it tries to inject the label into the kubelet configuration. the latter would not be what i would advocate for.

**Comment by javanthropus** [comment]
The ability to have karpenter set restricted labels would be helpful and makes administrative sense.  Those labels are restricted in order to prevent a subverted node from updating its own labels in order to attract new pods to itself.  Without an ability to set such labels as an admin, I'm forced to use unrestricted labels instead, and as a result a subverted node could set such labels for itself in order to attract sensitive pods.  The pods, their secrets, and their service accounts' access permissions would all be vulnerable at that point since the attacker would have direct access to all of them.

**Comment by sftim** [comment]
@elmiko have you got time to revise the issue description?

AIUI `node-role.kubernetes.io/worker` is official; when we added the code to `kubectl` and `kubeadm`, we missed updating the docs correctly to show this. So it's not that the labels are not well known; it's that they are not on the allow list for (official) keys that the kubelet is allowed to set on its own node.

**Comment by elmiko** [comment]
sure thing @sftim , were you thinking i should scope it tighter around the `node-role.kubernetes.io/worker` and remove mentions of the well-known labels in favor of things that should be on the allow list, or did you have something else in mind?

**Comment by elmiko** [comment]
@javanthropus thanks, i tend to agree with you in that i think this covers a larger group of interests than just my own, or the one label. imo, it covers the behavior pattern, if karpenter is adding these labels after the node is created, then should we stop it from adding anything?

**Comment by javanthropus** [comment]
I don't see a technical reason to stop it from adding arbitrary labels.  In fact, maybe it's even a good idea for karpenter to enforce that all labels are always applied throughout the lifetime of a node in order to avoid drift on those labels should an administrator try to change one for an arbitrary node.  All karpenter would need to do to keep things simple is filter out the restricted labels from those given to kubelet so that kubelet doesn't complain about them.

**Comment by sftim** [comment]
> were you thinking i should scope it tighter around the `node-role.kubernetes.io/worker` and remove mentions of the well-known labels in favor of things that should be on the allow list, or did you have something else in mind?

I think the issue is now about allowing Karpenter to manage [all] node labels that the kubelet is denied from setting.

**Comment by elmiko** [comment]
thanks, i'll spend some time rewriting it.

**Comment by sftim** [comment]
As a side note to this issue: I'd love to see a Boolean input to the Helm charts (eg for the AWS Karpenter integration), where you have to opt in to the RBAC binding that lets Karpenter manage node labels directly.

AIUI Karpenter doesn't need any access to write to nodes in its default configuration, and if we're able to keep that the default, it's a good outcome for security.

**Comment by elmiko** [comment]
that sounds like a good idea to me

**Comment by elmiko** [comment]
@sftim changed the titled and added an extra sentence about using any label. does that look good to you?

**Comment by sftim** [comment]
I suggest dropping or rewording this bit:
> As an openshift user, i require new nodes to take the label node-role.kubernetes.io/worker. Although this label uses the kubernetes.io prefix, it is not an [official well-known label](https://kubernetes.io/docs/reference/labels-annotations-taints/). Currently, karpenter does not allow NodePools to be created when a label with the kubernetes.io prefix is used on a non-well-known label for new node creation, or when checking predicates.

**Comment by enxebre** [comment]
node-role.kubernetes.io/ namespace was originally used and intended for conventional and organizational purposes. As captured in [1] It was not meant for core kube internal contractual consumption but there's nothing preventing their usage for consumers - "The use of the node-role.kubernetes/* label namespace is reserved solely for end-user and external Kubernetes consumers".

I see managing kubelet labels as task owned by the Service Provider Persona via the the bootstrap userdata implementation and the node restriction admission plugin.
The security boundary is already handled by the node restriction admission plugin. I wouldn't expect karpenter to duplicate this logic, potentially inconsistently drifting over time, specially since karpenter supports N-M kube versions.
Any opinions and earlier validation are specific to and should be owned by the bootstrap userdata implementation.

I'm not quite seeing how Karpenter preventing the usage of node-role.kubernetes.io selector will really harden security given  that as mentioned above any invalid input for kubelet labels will be stopped by the node restriction admission if the bootstrap userdata implementation tries to pass it through via kubelet. Besides, the node restriction admission plugin already "provide a place under the kubernetes.io label namespace for kubelets to self-label with kubelet and node-specific labels" [2] so any workload can actually use that allowed selector and it'll pass through via kubelet labels.

I see hardened workload isolation as a Cluster Admin / Service Provider persona task. They can rely on the node-restriction.kubernetes.io/* label prefix to label their Node objects centrally for isolation purposes.

Overall based on the above, I'd expect core karpenter to not make any imposition, and let the bootstrap userdata implementations to impose their opinions and earlier validations for labels.

[1] https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/1143-node-role-labels/README.md#goals

[2] https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/279-limit-node-access/README.md#proposal

**Comment by sftim** [comment]
Karpenter opted to pick the narrower option until the situation was cleared up. That's a sound choice if there's a worry that the other option could lead to a problem. Imagine if it turned out there was a risk but Karpenter users were already relying on the problem behavior.

Now what's needed is volunteer time to get things sorted, including the Kubernetes docs (so that it's clear to reviewers what is truly expected).

**Comment by enxebre** [comment]
I put a PR and tried to clarify in the description https://github.com/kubernetes-sigs/karpenter/pull/1908

**Comment by frimik** [comment]
> I'm not quite seeing how Karpenter preventing the usage of node-role.kubernetes.io selector will really harden security given that as mentioned above any invalid input for kubelet labels will be stopped by the node restriction admission if the bootstrap userdata implementation tries to pass it through via kubelet. Besides, the node restriction admission plugin already "provide a place under the kubernetes.io label namespace for kubelets to self-label with kubelet and node-specific labels" [2] so any workload can actually use that allowed selector and it'll pass through via kubelet labels.

Am I interpreting this and the related provider discussions that you've now reached the understanding that:

- kubelet ("the node") has restrictions on which labels **it** can set. Fine. Shouldn't let that influence _other_ avenues for label setting.
- Karpenter should be able to follow the operators intentions

If so, this would put Karpenter back on par for us with the simple labelling capabilities we had via Managed Node Groups. 👍 Thanks for seeing this through. @enxebre is revisiting #1908 on your calendar soon? 🙏

**Comment by enxebre** [comment]
sorry I got distracted fro this effort. I'll rebase and revive the PR soonish

**Comment by skoef** [comment]
@enxebre could you consider proceeding on that PR please? 🙏

**Comment by elmiko** [comment]
@skoef the PR has been proceeding under #2586 . we currently need reviews from approvers.
